src/settlement/worker.rs

```diff
@@ -205,4 +212,176 @@ impl Settler {
+                            let status_key = format!("{}{}", std::str::from_utf8(prefix::PREFIX_BLOCK_STATUS).unwrap(), last_submitted_block + 1);
+                            let status = Status::Submitted;
+                            let encoded_status = serde_json::to_vec(&status).unwrap();
+                            db.put(status_key.as_bytes().to_vec(), encoded_status);
+                        }
+                        Err(e) => {
+                            log::error!("submit block({}) failed: {:?}", last_submitted_block + 1, e);
+                        }
+                    }
+                }
+                _ = stop_rx.recv() => {
+                    log::info!("Submit Worker stopped");
+                    return Ok(());
+                }
+            }
+        }
+    }
+}
+
+fn convert_to_tx_legacy(tx: &Transaction) -> Result<TxLegacy> {
+    // create a legacy transaction
+    let chain_id = tx.chain_id.ok_or_else(|| anyhow!("chain id is required"))?;
+    let gas_price = tx
+        .gas_price
+        .ok_or_else(|| anyhow!("gas price is required"))?;
+    let input = Bytes::from(tx.input.clone().to_vec());
+
+    let tx_legacy = TxLegacy {
+        chain_id: Some(chain_id.as_u64()),
+        nonce: tx.nonce.as_u64(),
+        gas_price: gas_price.as_u128(),
+        gas_limit: tx.gas.as_u64(),
+        to: match tx.to {
+            Some(address) => {
+                TransactionKind::Call(reth_primitives::Address::from_slice(address.as_bytes()))
+            }
+            None => TransactionKind::Create,
+        },
+        value: reth_primitives::alloy_primitives::Uint::from(tx.value.as_u128()),
+        input,
+    };
+
+    Ok(tx_legacy)
+}
+
+// === wrap the reth/crates/primitives/src/transaction/legacy.rs TxLegacy private methods ===
+
+pub fn payload_len_for_signature(tx: &TxLegacy) -> usize {
+    let payload_length = fields_len(tx) + eip155_fields_len(tx);
+    // 'header length' + 'payload length'
+    length_of_length(payload_length) + payload_length
+}
+
+pub fn encode_for_signing(tx: &TxLegacy, out: &mut dyn bytes::BufMut) {
+    let payload_length = fields_len(tx) + eip155_fields_len(tx);
+    Header {
+        list: true,
+        payload_length,
+    }
+    .encode(out);
+    encode_fields(tx, out);
+    encode_eip155_fields(tx, out);
+}
+
+pub fn fields_len(tx: &TxLegacy) -> usize {
+    tx.nonce.length()
+        + tx.gas_price.length()
+        + tx.gas_limit.length()
+        + tx.to.length()
+        + tx.value.length()
+        + tx.input.0.length()
+}
+
+pub fn eip155_fields_len(tx: &TxLegacy) -> usize {
+    if let Some(id) = tx.chain_id {
+        id.length() + 2
+    } else {
+        0
+    }
+}
+
+pub fn encode_fields(tx: &TxLegacy, out: &mut dyn bytes::BufMut) {
+    tx.nonce.encode(out);
+    tx.gas_price.encode(out);
+    tx.gas_limit.encode(out);
+    tx.to.encode(out);
+    tx.value.encode(out);
+    tx.input.0.encode(out);
```

Comment:
Why do we encode input 0 only?

Reply:
input is a tuple struct that wraps a Bytes type.  `pub struct Bytes(pub bytes::Bytes);`
input.0 contains all data of the input field.